frontend/src/components/VideoChat.tsx

```diff
@@ -0,0 +1,362 @@
+      .then(stream => {
+        this.setState({ stream: stream });
+        const node = this.videoRef.current;
+        //TODO: Include id in this emit
+        this.socket.emit("newVideoChatPeer", this.state.videoChatId); //handled in server.js
+        if (node) {
+          node.srcObject = stream;
+        }
+
+        this.setState({
+          videoTrack: stream.getTracks()[1],
+          audioTrack: stream.getTracks()[0]
+        });
+      })
+      .catch(() => console.log("Permission Not Given")); //If permission not given, display error
+  };
+
+  stopMyVideoChat = () => {
+    this.state.videoTrack?.stop();
+    this.state.audioTrack?.stop();
+    this.setState({
+      videoTrack: null,
+      audioTrack: null,
+      inVideoChat: false,
+      peerConnected: false,
+      gotAnswer: false,
+      stream: undefined,
+      peer: null,
+      peerVideo: null,
+      videoChatSet: [],
+      videoChatId: null,
+      openInviteModal: false,
+      inviteFrom: ""
+    });
+    //TODO: SEND Leave video chat event to other use in video chat
+    //On receipt, they should stop stopVideoChat
+  };
+
+  setVideoChatId = (id: any) => {
+    this.setState({ videoChatId: id });
+    this.socket.emit("VIDEO_CHAT", this.state.videoChatId);
+    this.setInVideoChat();
+  };
+
+  sendVideoChatId = (accept: any) => {
+    const videoChatId = uuidv1();
+    const videoChatIdObj = { id: videoChatId, receiver: accept.sender };
+    this.setVideoChatId(videoChatIdObj.id);
+    this.socket.emit("SEND_VIDEOCHATID", videoChatIdObj);
+  };
+
+  openInviteModal = (invite: any) => {
+    console.log("inviteFrom: " + invite.sender);
+    this.setState({ inviteFrom: invite.sender, inviteFromName: invite.name });
+    this.setState({ openInviteModal: true });
+  };
+
+  sendInvite = (receiverId: string) => {
+    const name = this.state.name && this.state.name;
+    const invite = { sender: this.socket.id, receiver: receiverId, name: name };
+    this.socket.emit("SEND_INVITE", invite);
+  };
+
+  setVideoChatSet = (arr: string[]) => {
+    this.setState({ videoChatSet: arr });
+  };
+
+  acceptVideoChatInvite = () => {
+    const accept = { sender: this.socket.id, receiver: this.state.inviteFrom };
+    this.socket.emit("ACCEPT_INVITE", accept);
+    this.setState({ openInviteModal: false });
+  };
+
+  declineVideoChatInvite = () => {
+    //TODO: Send an event that says declined?
+    this.setState({ openInviteModal: false });
+  };
+
+  //Used to initialize a peer, define a new peer and return it
+  init = (type: PeerTypes): Peer.Instance => {
+    //type tells us if init == true means sends offer, if init is false, wont send offer, wait for offer, send answer
+    const socket = this.socket;
+    const peer = new Peer({
+      initiator: type === PeerTypes.init ? true : false,
+      stream: this.state.stream && this.state.stream,
+      trickle: false
+    });
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    const copyThis = this;
```

Comment:
what's this for?

Reply:
If you're trying to preserve the component's `this` inside callbacks then you should use arrow functions. Arrow functions were made so that they don't have their own bindings of `this` and just use the present context's `this` (the component itself here).